Tee-Object: add `-Encoding` parameter

rpolley opened the pull request:
# PR Summary

Fix #11104.

Adds a new -Encoding parameter to `Tee-Object` that controls what text encoding to use for the output file. If not provided, the cmdlet behaves as before and uses the default encoding.

## PR Context

Most cmdlets that deal with text file IO have an `-Encoding` parameter that tells the cmdlet which encoding to use when reading/writing the file. However, `Tee-Object` does not. This pull request adds this functionality. Closes #11104.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/pull/7902
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

rpolley committed:
Add encoding parameter to Tee-Object

rpolley committed:
Add encoding parameter to Tee-Object

rpolley committed:
Merge branch 'Add--Encoding-parameter-for-Tee-Object' of github.com:rpolley/PowerShell into Add--Encoding-parameter-for-Tee-Object

rpolley committed:
Merge branch 'Add--Encoding-parameter-for-Tee-Object' of github.com:rpolley/PowerShell into Add--Encoding-parameter-for-Tee-Object

rpolley committed:
Merge branch 'Add--Encoding-parameter-for-Tee-Object' of github.com:rpolley/PowerShell into Add--Encoding-parameter-for-Tee-Object

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1:
```diff
@@ -18,7 +18,28 @@ Describe "Tee-Object" -Tags "CI" {
-	    }
+        }
+
+        It "Should tee output to file using selected encoding" {
+            # Pass Encoding to Tee-Object
+            # and then read the resulting object back
+            # using the selected encoding
+            ForEach($encoding in @(
+                "ascii",
```
Please remove obvious comments - we already have the `It` title.
Also please use `-TestCase` parameter instead of `ForEach`. (See examples in other tests).

rpolley replied:
This is fixed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1:
```diff
@@ -18,7 +18,28 @@ Describe "Tee-Object" -Tags "CI" {
+                "utf8"
+            )){
+                $teefile = $testfile
+	            Write-Output teeobjecttest3  | Tee-Object $teefile -Encoding $encoding
+	            Get-Content $teefile -Encoding $encoding | Should -BeExactly "teeobjecttest3"
+	            Remove-Item $teefile -ErrorAction SilentlyContinue
```
Please use named parameters in cmdlets.

rpolley replied:
This is fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs:
```diff
@@ -85,6 +86,22 @@ public string Variable
+        [ArgumentToEncodingTransformationAttribute]
+        [ArgumentEncodingCompletionsAttribute]
+        [ValidateNotNullOrEmpty]
+        public Encoding Encoding
+        {
+            get { return _encoding; }
+
+            set { _encoding = value; }
+        }
+
+        private Encoding _encoding = ClrFacade.GetDefaultEncoding();
```
Please use auto-implemented property:
```suggestion
        public Encoding Encoding { get; set; } = ClrFacade.GetDefaultEncoding();
```

rpolley replied:
This is fixed.

iSazonov reviewed (commented):
Please fix CIs.

rpolley committed:
clean up test case

rpolley committed:
use auto-implemented property declaration

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs:
```diff
@@ -85,6 +86,15 @@ public string Variable
+        /// <summary>
+        /// Sets the encoding to use when writing the file.
+        /// </summary>
+        [Parameter]
+        [ArgumentToEncodingTransformationAttribute]
+        [ArgumentEncodingCompletionsAttribute]
+        [ValidateNotNullOrEmpty]
+        public Encoding Encoding { get; set; } = ClrFacade.GetDefaultEncoding();
```
I guess CIs fail because we need explicitly make the parameter optional (Mandatory = false).

iSazonov reviewed (commented):
@rpolley Please look test fails.

rpolley committed:
explicitly set mandatory=false

iSazonov commented on \:
```diff
@@ -0,0 +1,14 @@
+Add encoding parameter to Tee-Object
+
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+#
+# Date:      Mon Jan 4 18:27:32 2021 -0600
+#
+# On branch Add--Encoding-parameter-for-Tee-Object
+# Your branch is up to date with 'origin/Add--Encoding-parameter-for-Tee-Object'.
+#
+# Changes to be committed:
+#	modified:   src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs
+#	modified:   test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1
+#
```
Please remove the extra file - build process fails on it.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs:
```diff
@@ -85,6 +86,15 @@ public string Variable
+        /// <summary>
+        /// Sets the encoding to use when writing the file.
+        /// </summary>
+        [Parameter(Mandatory = false)]
```
Please revert - I was wrong.

rpolley committed:
Delete \

Looks like this was some sort of temp file created during merge commit that ended up in the branch somehow

rpolley committed:
Remove mandatory = false from parameter annotation

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs:
```diff
@@ -110,6 +120,9 @@ protected override void BeginProcessing()
             }
+            
+            // passthrough Encoding parameter
+            _commandWrapper.AddNamedParameter("Encoding", Encoding);
```
This doesn't work for ` Tee-Object -Variable` because we assign Encoding parameter to Set-Variable cmdlet. Please see previous https://github.com/PowerShell/PowerShell/blob/b303c2f614b8e158179ef9e2c2fe73110dbfb654/src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs#L117-L121

rpolley replied:
Didn't even notice this, thanks. I updated it so that the Encoding parameter is only used for the `File` and `LiteralFile` parameter sets.

rpolley committed:
confine Encoding parameter to File parameter sets

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Tee-Object.cs:
```diff
@@ -85,6 +86,16 @@ public string Variable
         private string _variable;
 
+        /// <summary>
+        /// Sets the encoding to use when writing the file.
```
```suggestion
        /// Gets and sets the encoding to use when writing the file.
```

rpolley replied:
Applied the suggestion, but it seems like that didn't fix the codeclimate issue.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Tee-Object.Tests.ps1:
```diff
@@ -18,7 +18,24 @@ Describe "Tee-Object" -Tags "CI" {
+        ) {
+            param($Encoding)
+            $teefile = $testfile
+            Write-Output -InputObject teeobjecttest3  | Tee-Object -FilePath $teefile -Encoding $Encoding
```
Really the test doesn't check encoding because write ASCII only.
Please see `Verifies -Tail with content that uses an explicit/implicit encoding` from Get-Content.Tests.ps1 . We could use the same string with native chars instead of `teeobjecttest3`.

rpolley committed:
parameter documentation is tied to the Parameter annotation?

rpolley committed:
Revert "parameter documentation is tied to the Parameter annotation?"

This reverts commit 3890059dd877cb9bb8122886482f30ef76a903e0.

rpolley committed:
Update Encoding documentation

Co-authored-by: Ilya <darpa@yandex.ru>

rpolley committed:
use unicode chars in encoding test

iSazonov commented:
@rpolley Please look CodeFactor and CI issues.

rpolley committed:
fix tee-object encoding test

rpolley committed:
use "Gets or Sets" in param documentation

rpolley committed:
fix documentation capitalization

daxian-dbw committed:
Merge branch 'master' into Add--Encoding-parameter-for-Tee-Object

daxian-dbw committed:
Resolve conflicts

iSazonov commented:
We need to open doc issue.

sdwheeler commented:
@iSazonov This was documented in https://github.com/MicrosoftDocs/PowerShell-Docs/pull/7902
